native_client/deepspeech.cc

```diff
@@ -8,45 +8,423 @@
+                          const short* buffer,
+                          unsigned int buffer_size)
+{
+  while (buffer_size > 0) {
+    while (buffer_size > 0 && ctx->audio_buffer_len < AUDIO_WIN_LEN_SAMPLES) {
+      ctx->audio_buffer[ctx->audio_buffer_len] = *buffer;
+      ++ctx->audio_buffer_len;
+      ++buffer;
+      --buffer_size;
+    }
+
+    if (ctx->audio_buffer_len == AUDIO_WIN_LEN_SAMPLES) {
+      processAudioWindow(ctx, ctx->audio_buffer, ctx->audio_buffer_len);
+      // shift data by one step of 10ms
+      memmove(ctx->audio_buffer, ctx->audio_buffer + AUDIO_WIN_STEP_SAMPLES, (AUDIO_WIN_LEN_SAMPLES - AUDIO_WIN_STEP_SAMPLES) * sizeof(short));
+      ctx->audio_buffer_len -= AUDIO_WIN_STEP_SAMPLES;
+    }
+  }
+}
+
+char*
+Private::finishStream(StreamingState* ctx)
+{
+  // flush audio buffer
+  processAudioWindow(ctx, ctx->audio_buffer, ctx->audio_buffer_len);
+
+  // add empty mfcc vectors at end of sample
+  for (int i = 0; i < MFCC_CONTEXT; ++i) {
+    addZeroMfccWindow(ctx);
+  }
+
+  // process final batch
+  if (ctx->batch_buffer_len > 0) {
+    processBatch(ctx, ctx->batch_buffer, ctx->batch_buffer_len/MFCC_FEATS_PER_TIMESTEP);
+  }
+
+  char* str = decode(ctx->accumulated_logits);
+  delete ctx;
+  return str;
+}
+
+void
+Private::processAudioWindow(StreamingState* ctx, short* buf, unsigned int len)
+{
+  ctx->skip_next_mfcc = !ctx->skip_next_mfcc;
+  if (!ctx->skip_next_mfcc) { // was true
+    return;
+  }
+
+  // Compute MFCC features
+  float* mfcc;
+  int n_frames = csf_mfcc(buf, len, SAMPLE_RATE,
+                          AUDIO_WIN_LEN, AUDIO_WIN_STEP, MFCC_FEATURES, N_FILTERS, N_FFT,
+                          LOWFREQ, SAMPLE_RATE/2, COEFF, CEP_LIFTER, 1, nullptr,
+                          &mfcc);
+  assert(n_frames == 1);
+
+  float* buffer = mfcc;
+  unsigned int bufferSize = n_frames * MFCC_FEATURES;
+  while (bufferSize > 0) {
+    while (bufferSize > 0 && ctx->mfcc_buffer_len < MFCC_FEATS_PER_TIMESTEP) {
+      ctx->mfcc_buffer[ctx->mfcc_buffer_len] = *buffer;
+      ++ctx->mfcc_buffer_len;
+      ++buffer;
+      --bufferSize;
+    }
+
+    if (ctx->mfcc_buffer_len == MFCC_FEATS_PER_TIMESTEP) {
+      processMfccWindow(ctx, ctx->mfcc_buffer, ctx->mfcc_buffer_len);
+      // shift data by one step of one mfcc feature vector
+      memmove(ctx->mfcc_buffer, ctx->mfcc_buffer + MFCC_FEATURES, (ctx->mfcc_buffer_len - MFCC_FEATURES) * sizeof(float));
+      ctx->mfcc_buffer_len -= MFCC_FEATURES;
+    }
+  }
+
+  free(mfcc);
+}
+
+void
+Private::processMfccWindow(StreamingState* ctx, float* buf, unsigned int len)
+{
+  while (len > 0) {
+    while (len > 0 && ctx->batch_buffer_len < N_STEPS_PER_BATCH*MFCC_FEATS_PER_TIMESTEP) {
+      ctx->batch_buffer[ctx->batch_buffer_len] = *buf;
+      ++ctx->batch_buffer_len;
+      ++buf;
+      --len;
+    }
+
+    if (ctx->batch_buffer_len == N_STEPS_PER_BATCH*MFCC_FEATS_PER_TIMESTEP) {
+      processBatch(ctx, ctx->batch_buffer, N_STEPS_PER_BATCH);
```

Comment:
I'm a bit lost with `processBatch` in the middle of `processMfccWindow`. i'll have to check that more closely

Reply:
`processMfccWindow` takes MFCC features, copies them to the batch buffer, and if the batch buffer is full, computes a batch in order to be able to clear it and continue processing MFCC features.